### PR TITLE
Adds strains framework

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request

To add a strain just add a child with a UPGRADE_BASETYPE upgrade to a castedatum has a flag for snowflake like weapon x

## Why It's Good For The Game

this intended for sidegrades within the same castes role

### CODE REVIEW

**Comment by yyzsong** [comment]
peak is here

File: code/modules/mob/living/carbon/xenomorph/evolution.dm [hunk 1]
```diff
@@ -16,14 +16,27 @@
...
-	if(world.time - (GLOB.key_to_time_of_caste_swap[key] ? GLOB.key_to_time_of_caste_swap[key] : -INFINITY) < 9000) //casteswap timer, 15 minutes
+	if(world.time - (GLOB.key_to_time_of_caste_swap[key] ? GLOB.key_to_time_of_caste_swap[key] : -INFINITY) < (15 MINUTES))
 		to_chat(src, span_warning("Your caste swap timer is not done yet."))
 		return
 
 	SStgui.close_user_uis(src, GLOB.evo_panel)
 	ADD_TRAIT(src, TRAIT_CASTE_SWAP, TRAIT_CASTE_SWAP)
 	GLOB.evo_panel.ui_interact(src)
 
+/mob/living/carbon/xenomorph/verb/strain_swap()
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/evolution.dm:27]: Verb ?

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 2]
```diff
@@ -245,6 +245,21 @@
 		current_type = initial(current_type.parent_type)
 	return current_type
 
+///returns a list of strains(xeno castedatum paths) that this caste can currently evolve to
+/datum/xeno_caste/proc/get_strain_options()
+	var/datum/xeno_caste/root_type = type
+	while(initial(root_type.parent_type) != /datum/xeno_caste)
+		root_type = initial(root_type.parent_type)
+	var/list/candidates = typesof(root_type)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:253]: Why do you have to do this ?

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/evolution.dm (hunk 1 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/evolution.dm:27]: We still dont have an option for this, and caste swap/evo/regress are the same so

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm (hunk 2 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:253]: We get the child of xeno_caste (e.g runner)  then get all children of that to find the strains

Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:253]: Why not a Glob list somewhere ? less snowflaky

Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:253]: How so? I need to return the strains for a given type, and I dont want to force people to add defines to everything for a glob lookup

Comment (Lumipharon) [reply code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:253]: Can't you just have a GLOBAL_LIST_INIT to do this?

**Review by Lumipharon — CHANGES_REQUESTED** [review]
Rest looks good otherwise.

As I've brought up before though, I still don't understand  codewise what the difference is between something being a strain vs just being a new caste.

We no longer have evo paths so we can evo to everything, and can swap between castes at will (with changing strain literally using the same timer)... So other than organising them in the hive UI, wouldn't it just be easier to treat them as new castes? 

Like it can still be caste/runner/snowflake or whatever, but you just evo directly to it, instead of evo'ing runner, then swapping to snowflake strain.

All this is really doing is adding evo paths back in a more limited fashion.

**Comment by lKiseki** [comment]
> I'm happy to merge if you disagree, but this seems to just be a more complicated way to get the exact same result.

Making new castes is nott always a good solution, and you can see why if you compare the Crusher to the Bull. The latter could easily be a strain of the former, especially because it just doesn't work as a standalone caste. It could be fun and interesting to design minor variations of existing castes that offer slightly different gameplay experiences. Having to make an entirely new thing, on the other hand, is much more difficult by comparison. I would very much prefer having strains.

**Comment by Lumipharon** [comment]
> > I'm happy to merge if you disagree, but this seems to just be a more complicated way to get the exact same result.
> 
> Making new castes is not always a good solution, and you can see why if you compare the Crusher to the Bull. The latter could easily be a strain of the former, especially because it just doesn't work as a standalone caste. It could be fun and interesting to design minor variations of existing castes that offer slightly different gameplay experiences. Having to make an entirely new thing, on the other hand, is much more difficult by comparison. I would very much prefer having strains.

This has nothing to do with the ingame stats/abilities/appearance of a strain. I don't know why people seem confused by this every time I bring it up. Nothing I am saying is 'don't do strains'. 
I am saying any xeno created as a strain is codewise just a new caste, and ingame the only difference between a new caste and a new strain is the strain has an arbitrary extra step to get to (i.e. you gotta evo/caste swap to the base caste THEN strain swap to the strain), which is literally just readding evo paths but more limited.

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm (hunk 2 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:253]: Done

**Comment by TiviPlus** [comment]
> As I've brought up before though, I still don't understand codewise what the difference is between something being a strain vs just being a new caste.

Codewise its the same yea
Point being is at the end of the day we can have the same role with multiple flavors. e.g different boiler is siege bombarder but you can have a few tweaks to tailor to precise player playtypes

**Comment by Lumipharon** [comment]
ahh, I feel like I have brain damage here, I don't know why what I am saying is not clear.

The end result is exactly the same. I made weapon X for campaign, this could be described as a strain of hunter, since its literally just hunter with a few buffs.
Codewise its simply a new caste though.

No strain code was needed to make it. You can add in acid runner, hedgehog rav, all these strain things as new castes, or as strains **and they are 100% the same either way.** Being a strain or being a caste has zero bearing on what it looks like, or its similarity to another caste in terms of abilities or stats or whatever.

The only difference between adding acid runner as a 'strain' codewise vs a new caste codewise, is we have readded evolutionary paths as a game mechanic, immediately after we removed evolutionary paths as a game mechanic for being bad.

The acid runner is still going to be xeno_caste/runner/acid_runner regardless of the whether its a strain or a caste. its still going to have the same sprite. The same stats. The same abilities.
The only thing different is if its a strain, I have to evo to runner, THEN have a timer to evo to acid runner.
Vs if its a caste, I just evo/swap directly to acid runner.

Why are we readding this mechanic if we just removed it?

**Comment by lKiseki** [comment]
> All I am saying is why require players to evolve or swap to runner, before being able to strain swap to acid runner or whatever? **Just treat the strains how we treat castes**. Larva evolves directly to acid runner, or hivelord devolves directly to acid runner, or defender just swaps directly to acid runner etc.

If what you're referring to is "why not add the option to evo straight to a strain from the evo menu", it's probably because that's convoluted and would either extend the list to be really long, or require a redesign of the TGUI interface, probably with collapsible lists per caste.

**Comment by Lumipharon** [comment]
The argument for adding an entire extra mechanic (that shares a time gate with **caste** swapping) due to not having enough space in a window that can be any size is uh, highly questionable.

**Review by Lumipharon — APPROVED** [review]

**Comment by PibeCantina** [comment]
Merge when

**Comment by chizzy376** [comment]
> Merge when

soon tm

**Comment by hivemuscle** [comment]
will we have cqc boiler finally?

**Pull request merged**